tensorboard/uploader/formatters.py

```diff
@@ -0,0 +1,100 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Helpers that format the information about experiments as strings."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import collections
+import json
+
+import six
+
+from tensorboard.uploader import util
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseExperimentFormatter(object):
+    """Abstract base class for formatting experiment information as a string."""
+
+    @abc.abstractmethod
+    def format_experiment(self, experiment, experiment_url):
+        """Format the information about an experiment as a representing string.
+
+        Args:
+          experiment: An `experiment_pb2.Experiment` protobuf message for the
+            experiment to be formatted.
+          experiment_url: The URL at which the experiment can be accessed via
+            TensorBoard.
+
+        Returns:
+          A string that represents the experiment.
+        """
+        pass
+
+
+class ReadableFormatter(BaseExperimentFormatter):
+    """A formatter implementation that outputs human-readable text."""
+
+    _NAME_COLUMN_WIDTH = 12
+
+    def __init__(self):
+        super(ReadableFormatter, self).__init__()
+
+    def format_experiment(self, experiment, experiment_url):
+        output = []
+        output.append(experiment_url)
+        data = [
+            ("Name", experiment.name or "[No Name]"),
+            ("Description", experiment.description or "[No Description]"),
+            ("Id", experiment.experiment_id),
+            ("Created", util.format_time(experiment.create_time)),
+            ("Updated", util.format_time(experiment.update_time)),
+            ("Runs", str(experiment.num_runs)),
+            ("Tags", str(experiment.num_tags)),
+            ("Scalars", str(experiment.num_scalars)),
+        ]
+        for name, value in data:
+            output.append(
+                "\t%s %s" % (name.ljust(self._NAME_COLUMN_WIDTH), value,)
+            )
+        return "\n".join(output)
+
+
+class JsonFormatter(object):
+    """A formatter implementation: outputs experiment as JSON."""
+
+    _JSON_INDENT = 2
+
+    def __init__(self):
+        super(JsonFormatter, self).__init__()
+
+    def format_experiment(self, experiment, experiment_url):
+        data = [
+            ("url", experiment_url),
+            ("name", experiment.name),
+            ("description", experiment.description),
+            ("id", experiment.experiment_id),
+            ("created", util.format_time(experiment.create_time)),
+            ("updated", util.format_time(experiment.update_time)),
```

Comment:
It probably makes more sense for to use `format_time_absolute` for the
JSON formatter? Since the point is to have a simple, parseable format,
outputs like `"just now"` probably aren’t what’s wanted.

Reply:
Done. Unit test is updated accordingly.